doc/contributing/erm-guidelines.md

```diff
@@ -0,0 +1,307 @@
+successful and exceptional exits from the scopes in which the using keyword
+is used. This means that if an exception is thrown within the scope, the
+disposal methods will still be called. However, when the disposal methods are
+called they are not aware of the context in which they were called. These
+methods will not receive any information about the exception that was thrown
+or whether an exception was thrown at all. This means that it is safest to
+assume that the disposal methods will be called in a context where the object
+may not be in a valid state or that an exception may be pending.
+
+## Guidelines for Disposable Objects
+
+So with this is mind, it is necessary to outline some guidelines for disposers:
+
+1. Disposers should be idempotent. Multiple calls to the disposal methods
+   should not cause any issues or have any additional side effects.
+2. Disposers should assume that it is being called in an exception context.
+   Always assume there is likely a pending exception and that if the object
+   has not been explicitly closed when the disposal method is called, that
+   the object should be disposed as if an exception had occurred. For instance,
+   if the object API exposes both a `close()` method and an `abort()` method,
+   the disposal method should call `abort()` if the object is not already
+   closed. If there is no difference in disposing in success or exception
+   contexts, then separate disposal methods are unnecessary.
+3. Disposers may throw their own exceptions but this is not recommended.
+   If a disposer throws an exception while there is another pending
+   exception, then both exceptions will be wrapped in a `SupressedError`
+   that masks both. This makes it difficult to understand the context
+   in which the exceptions were thrown. It is, however, not possible to
+   completely prevent exceptions from being thrown in the disposal methods
+   so this guideline is more of a recommendation than a hard rule.
+4. Disposable objects should expose explicit disposal methods in addition
+   to the `Symbol.dispose` and `Symbol.asyncDispose` methods. This allows
+   user code to explicitly dispose of the object without using the `using`
+   or `await using` statements. For example, a disposable object might
+   expose a `close()` method that can be called to dispose of the object.
+   The `Symbol.dispose` and `Symbol.asyncDispose` methods should delegate to
+   these explicit disposal methods.
+5. Because it is safest to assume that the disposal method will be called
+   in an exception context, it is strongly recommended to just generally avoid
+   use of `Symbol.asyncDispose` as much as possible. Asynchronous disposal can
+   lead delaying the handling of exceptions and can make it difficult to
+   reason about the state of the object while the disposal is in progress and
+   is often an anti-pattern. Disposal in an exception context should always
+   be synchronous and immediate.
```

Comment:
Does this rule apply to resources that are initialized asynchronously but can be synchronously disposed of?
For example, let's say, we can implement [`filehandle.close()`](https://nodejs.org/api/fs.html#filehandleclose) in sync (and sane) way internally. Should we prefer `Symbol.dispose` in that case? From userspace perspective, I'd expect a rule of thumb that it should be
```mjs
using handle = createDisposableResource();
// or
await using handle = await createDisposableResource();
// but usually not:
using handle = await createDisposableResource();
```